### PR TITLE
Compile with -safe-string (OCaml 4.06)

### PR DESCRIPTION
**Pull request by vasilisp**
Use immutable strings as much as possible. Some copying needed.

See discussion in #135.

### CODE REVIEW

File: src/extensions/deflatemod.ml [hunk 1]
```diff
@@ -104,15 +105,16 @@ let rec output oz f buf pos len  =
     let  (_, used_in, used_out) =
       try
         Zlib.deflate
-          oz.stream buf pos len oz.buf oz.pos oz.avail Zlib.Z_NO_FLUSH
+          oz.stream (Bytes.unsafe_of_string buf)
```
Comment (hhugo) [inline src/extensions/deflatemod.ml]: what about have buf be a `bytes`

File: src/http/ocsigen_http_com.ml [hunk 2]
```diff
@@ -586,14 +587,17 @@ let (<<?) h (n, v) =
   | Some v -> Http_headers.replace n v h
 
 let gmtdate d =
-  let x = Netdate.mk_mail_date ~zone:0 d in try
+  let x = Netdate.mk_mail_date ~zone:0 d in
+  try
+    let x = Bytes.unsafe_of_string x in
```
Comment (hhugo) [inline src/http/ocsigen_http_com.ml]: Given that we already do a copy with `Bytes.sub_string`, can't we afford an extra copy ?
Bytes.unsafe_of_string is hard to reason about. What if Netdate.mk_mail_date start to do memoization..

File: src/extensions/deflatemod.ml (hunk 1 above)
Comment (hhugo) [reply src/extensions/deflatemod.ml]: Alternatively, one could patch zlib to add strings to the API ..

Comment (balat) [reply src/extensions/deflatemod.ml]: I just did a PR for adding this in Zlib API :)

Comment (balat) [reply src/extensions/deflatemod.ml]: We tried the other version. It leads to lots of modifications in the code ...

**Pull request merged**